lib/redis_memo/memoize_query/cached_select/bind_params.rb

```diff
@@ -55,73 +115,68 @@ def product(other)
       #    {a: 2, b: 1},
       #    {a: 2, b: 2},
       #  ]
-      return unless other
-
-      # The tree is almost always right-heavy. Merge into the right node for better
-      # performance.
-      params.each do |model, attrs_set|
-        next if attrs_set.empty?
+      left.merge(right) do |_, attrs_set, other_attrs_set|
+        next attrs_set if other_attrs_set.empty?
+        next other_attrs_set if attrs_set.empty?
 
-        # The other model does not have any conditions so far: carry the
-        # attributes over to the other node
-        if other.params[model].empty?
-          other.params[model] = attrs_set
-          next
-        end
-
-        # Distribute the current attrs into the other
-        other_attrs_set_size = other.params[model].size
-        other_attrs_set = other.params[model]
-        merged_attrs_set = Array.new(other_attrs_set_size * attrs_set.size)
-
-        attrs_set.each_with_index do |attrs, i|
-          other_attrs_set.each_with_index do |other_attrs, j|
-            k = i * other_attrs_set_size + j
-            merged_attrs = merged_attrs_set[k] = other_attrs.dup
+        # distribute the current attrs into the other
+        merged_attrs_set = Set.new
+        attrs_set.each do |attrs|
+          other_attrs_set.each do |other_attrs|
+            merged_attrs = other_attrs.dup
+            should_add = true
             attrs.each do |name, val|
-              # Conflict detected. For example:
+              # conflict detected. for example:
               #
               #   (a = 1 or b = 1) and (a = 2 or b = 2)
               #
-              #  Keep:     a = 1 and b = 2, a = 2 and b = 1
-              #  Discard:  a = 1 and a = 2, b = 1 and b = 2
+              #  keep:     a = 1 and b = 2, a = 2 and b = 1
+              #  discard:  a = 1 and a = 2, b = 1 and b = 2
               if merged_attrs.include?(name) && merged_attrs[name] != val
-                merged_attrs_set[k] = nil
+                should_add = false
                 break
               end
 
               merged_attrs[name] = val
             end
+            merged_attrs_set << merged_attrs if should_add
           end
         end
 
-        merged_attrs_set.compact!
-        other.params[model] = merged_attrs_set
+        merged_attrs_set
       end
+    end
 
-      other
+    def plan_product
+      plan.dependency_size = left.plan.dependency_size * right.plan.dependency_size
+      plan.model_attrs = product_attrs_set(left.plan.model_attrs, right.plan.model_attrs)
```

Comment:
also here, i think `plan` can own the `product` method.

Reply:
similar to the above comment, product is used outside of the planning phase too